lib/fsm/utils_test.go

```diff
@@ -148,3 +149,248 @@ func (s *FSMUtilsSuite) TestDiffPlanNoPrevious(c *check.C) {
+					},
+					{
+						ID:       "/checks",
+						State:    storage.OperationPhaseStateUnstarted,
+						Requires: []string{"/startAgent"},
+					},
+				},
+			},
+			phaseID: "/init",
+		},
+		{
+			comment: "Rollback after a previously forced rollback",
+			plan: &storage.OperationPlan{
+				Phases: []storage.OperationPhase{
+					{
+						ID:    "/init",
+						State: storage.OperationPhaseStateCompleted,
+					},
+					{
+						ID:       "/startAgent",
+						State:    storage.OperationPhaseStateRolledBack,
+						Requires: []string{"/init"},
+					},
+					{
+						ID:       "/checks",
+						State:    storage.OperationPhaseStateFailed,
+						Requires: []string{"/startAgent"},
+					},
+				},
+			},
+			phaseID:  "/init",
+			expected: `rollback subsequent phases before rolling back phase "/init"`,
+		},
+		{
+			comment: "Rollback after a later phase has been executed out of band",
+			plan: &storage.OperationPlan{
+				Phases: []storage.OperationPhase{
+					{
+						ID:    "/init",
+						State: storage.OperationPhaseStateCompleted,
+					},
+					{
+						ID:       "/startAgent",
+						State:    storage.OperationPhaseStateUnstarted,
+						Requires: []string{"/init"},
+					},
+					{
+						ID:       "/checks",
+						State:    storage.OperationPhaseStateCompleted,
+						Requires: []string{"/startAgent"},
+					},
+					{
+						ID:       "/test",
+						State:    storage.OperationPhaseStateRolledBack,
+						Requires: []string{"/checks"},
+					},
+				},
+			},
+			phaseID:  "/init",
+			expected: `rollback subsequent phases before rolling back phase "/init"`,
+		},
+		{
+			comment: "Rollback subphase",
+			plan: &storage.OperationPlan{
+				Phases: []storage.OperationPhase{
+					{
+						ID:    "/masters",
+						State: storage.OperationPhaseStateInProgress,
+						Phases: []storage.OperationPhase{
+							{
+								ID:    "/masters/node-1",
+								State: storage.OperationPhaseStateCompleted,
+							},
+							{
+								ID:       "/masters/node-2",
+								State:    storage.OperationPhaseStateInProgress,
+								Requires: []string{"/masters/node-1"},
+							},
+						},
+					},
+				},
+			},
+			phaseID: "/masters/node-2",
+		},
+		{
+			comment: "Rollback non-leaf phase",
+			plan: &storage.OperationPlan{
+				Phases: []storage.OperationPhase{
+					{
+						ID:    "/masters",
+						State: storage.OperationPhaseStateCompleted,
+						Phases: []storage.OperationPhase{
+							{
+								ID:    "/masters/node-1",
+								State: storage.OperationPhaseStateCompleted,
+							},
+							{
+								ID:       "/masters/node-2",
+								State:    storage.OperationPhaseStateCompleted,
+								Requires: []string{"/masters/node-1"},
+							},
+						},
+					},
+				},
+			},
+			phaseID:  "/masters",
+			expected: "unable to rollback non-leaf phase, only leaf phases can be rolled back",
+		},
+		{
+			comment: "Top level phase has dependent phases that have not been rolled back",
+			plan: &storage.OperationPlan{
```

Comment:
Can you add a few more test cases with more complicated plans, i.e. multiple-level deep phases, multiple "requires", etc., more like our real plan. E.g.

```
/init
/masters <-- depends on /init
  * /node-1
    * /drain
    * /system-upgrade <-- depends on /masters/node-1/drain
  * /node-2
    * /drain
    * /system-upgrade <-- depends on /masters/node-2/drain
/nodes <-- depends on /masters
  * /node-3
    * /drai
    * /system-upgrade
/runtime <-- depends on /masters
  * /monitoring
  * /site
/gc <-- depends on /runtime
```

... you get the idea... and try different scenarios.